File: app/src/main/java/br/com/jonathanzanella/myexpenses/card/CardRepository.java
package br.com.jonathanzanella.myexpenses.card;

import android.support.annotation.NonNull;
import android.support.annotation.WorkerThread;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import java.util.List;
import java.util.UUID;

import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.database.Fields;
import br.com.jonathanzanella.myexpenses.database.ModelRepository;
import br.com.jonathanzanella.myexpenses.database.Repository;
import br.com.jonathanzanella.myexpenses.database.Where;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;
import br.com.jonathanzanella.myexpenses.validations.ValidationResult;

import static br.com.jonathanzanella.myexpenses.log.Log.warning;

public class CardRepository implements ModelRepository<Card> {
	private final Repository<Card> repository;
	private final ExpenseRepository expenseRepository;
	private final CardTable table = new CardTable();

	public CardRepository(Repository<Card> repository, ExpenseRepository expenseRepository) {
		this.repository = repository;
		this.expenseRepository = expenseRepository;
	}

	@WorkerThread
	public Card find(String uuid) {
		return repository.find(table, uuid);
	}

	@WorkerThread
	List<Card> all() {
		return repository.query(table, new Where(null).orderBy(Fields.NAME));
	}

	@WorkerThread
	public List<Card> creditCards() {
		return repository.query(table, new Where(Fields.TYPE).eq(CardType.CREDIT.getValue()));
	}

	@WorkerThread
	public Card accountDebitCard(Account account) {
		return repository.querySingle(table,
				new Where(Fields.TYPE)
						.eq(CardType.DEBIT.getValue())
						.and(Fields.ACCOUNT_UUID)
						.eq(account.getUuid()));
	}

	@WorkerThread
	public List<Card> unsync() {
		return repository.unsync(table);
	}

	@WorkerThread
	public long greaterUpdatedAt() {
		return repository.greaterUpdatedAt(table);
	}

	@WorkerThread
	public ValidationResult save(Card card) {
		ValidationResult result = validate(card);
		if(result.isValid()) {
			if(card.getId() == 0 && card.getUuid() == null)
				card.setUuid(UUID.randomUUID().toString());
			card.setSync(false);
			repository.saveAtDatabase(table, card);
		}
		return result;
	}

	@NonNull
	private ValidationResult validate(Card card) {
		ValidationResult result = new ValidationResult();
		if(StringUtils.isEmpty(card.getName()))
			result.addError(ValidationError.NAME);
		if(card.getType() == null)
			result.addError(ValidationError.CARD_TYPE);
		if(card.getAccount() == null)
			result.addError(ValidationError.ACCOUNT);
		return result;
	}

	@WorkerThread
	@Override
	public ValidationResult syncAndSave(final Card unsyncCard) {
		ValidationResult result = validate(unsyncCard);
		if(!result.isValid()) {
			warning("Card sync validation failed", unsyncCard.getData() + "\nerrors: " + result.getErrorsAsString());
			return result;
		}

		Card card = find(unsyncCard.getUuid());
		if(card != null && card.getId() != unsyncCard.getId()) {
			if(card.getUpdatedAt() != unsyncCard.getUpdatedAt())
				warning("Card overwritten", unsyncCard.getData());
			unsyncCard.setId(card.getId());
		}

		unsyncCard.setSync(true);
		repository.saveAtDatabase(table, unsyncCard);

		return result;
	}

	@WorkerThread
	public List<Expense> creditCardBills(Card card, DateTime month) {
		return expenseRepository.unpaidCardExpenses(month, card);
	}

	@WorkerThread
	public int getInvoiceValue(Card card, DateTime month) {
		int total = 0;
		for (Expense expense : creditCardBills(card, month))
			total += expense.getValue();

		return total;
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/card/CardRepository.kt
package br.com.jonathanzanella.myexpenses.card

import android.support.annotation.WorkerThread
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import br.com.jonathanzanella.myexpenses.validations.ValidationResult
import org.apache.commons.lang3.StringUtils
import org.joda.time.DateTime
import timber.log.Timber
import java.util.*

open class CardRepository(private val expenseRepository: ExpenseRepository) {

    @WorkerThread
    fun find(uuid: String): Card? {
        return App.database.cardDao().find(uuid).blockingFirst().firstOrNull()
    }

    @WorkerThread
    fun all(): List<Card> {
        return App.database.cardDao().all().blockingFirst()
    }

    @WorkerThread
    fun unsync(): List<Card> {
        return App.database.cardDao().unsync().blockingFirst()
    }

    @WorkerThread
    fun creditCards(): List<Card> {
        return App.database.cardDao().cards(CardType.CREDIT.value).blockingFirst()
    }

    @WorkerThread
    fun accountDebitCard(account: Account): Card? {
        return App.database.cardDao().accountCard(CardType.DEBIT.value, account.uuid!!).blockingFirst().firstOrNull()
    }

    @WorkerThread
    fun greaterUpdatedAt(): Long {
        return App.database.cardDao().greaterUpdatedAt().blockingFirst().firstOrNull()?.updatedAt ?: 0L
    }

    @WorkerThread
    fun save(card: Card): ValidationResult {
        val result = validate(card)
        if (result.isValid) {
            if (card.id == 0L && card.uuid == null)
                card.uuid = UUID.randomUUID().toString()
            card.sync = false
            card.id = App.database.cardDao().saveAtDatabase(card)
        }
        return result
    }

    private fun validate(card: Card): ValidationResult {
        val result = ValidationResult()
        if (StringUtils.isEmpty(card.name))
            result.addError(ValidationError.NAME)
        if (card.type == null)
            result.addError(ValidationError.CARD_TYPE)
        if (card.account == null)
            result.addError(ValidationError.ACCOUNT)
        return result
    }

    @WorkerThread
    fun syncAndSave(unsync: Card): ValidationResult {
        val result = validate(unsync)
        if (!result.isValid) {
            Timber.tag("Card sync valida failed")
                    .w(unsync.getData() + "\nerrors: " + result.errorsAsString)
            return result
        }

        val card = find(unsync.uuid!!)
        if (card != null && card.id != unsync.id) {
            if (card.updatedAt != unsync.updatedAt)
                Timber.tag("Card overwritten").w(unsync.getData())
            unsync.id = card.id
        }

        unsync.sync = true
        unsync.id = App.database.cardDao().saveAtDatabase(unsync)

        return result
    }

    @WorkerThread
    fun creditCardBills(card: Card, month: DateTime) =
        expenseRepository.unpaidCardExpenses(month, card)

    @WorkerThread
    fun getInvoiceValue(card: Card, month: DateTime) =
        creditCardBills(card, month).sumBy { it.value }
}
